pkg/cri/sbserver/nri-api.go

```diff
@@ -0,0 +1,781 @@
+//go:build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sbserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/pkg/cri/annotations"
+	cstore "github.com/containerd/containerd/pkg/cri/store/container"
+	sstore "github.com/containerd/containerd/pkg/cri/store/sandbox"
+	ctrdutil "github.com/containerd/containerd/pkg/cri/util"
+	"github.com/containerd/typeurl"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	runtimespec "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/generate"
+	cri "k8s.io/cri-api/pkg/apis/runtime/v1"
+
+	"github.com/containerd/containerd/pkg/cri/constants"
+	"github.com/containerd/containerd/pkg/nri"
+
+	"github.com/containerd/nri/pkg/api"
+
+	nrigen "github.com/containerd/nri/pkg/runtime-tools/generate"
+)
+
+type nriAPI struct {
```

Comment:
Possibly we can have `nriAPI` shared between `server` and `observer` ?
It mostly depends on sandboxStore object and a couple of CRI calls:

```go
type criDependencies interface {
	updateContainerResources() error
	config() criconfig.Config
}

type nriAPI struct {
	cri criDependencies // <-- plug server.criService or observer.criService
	sandboxStore *sandbox.Store
	nri nri.API
}
```

Reply:
Yes, actually the `nriAPI`s are now identical (the one in `sbserver` is a verbatim copy of the one in `server`). I was trying to think whether and how we could move nriAPI one level up in the tree to pkg/cri and use it in both, but I just drew a blank. I didn't think of that. I'll check if I can come up with something working based on your suggestion.

Reply:
@mxpv I split out the NRI integration bits to a common `pkg/cri/nri` package.

The implementation-specific bits are isolated into a `CRIImplementation` interface which is essentially your `criDependencies` above. I opted for making all the bits which that package needs from `sbserver/server` (CRI configuration, sandbox store, container store, even the cotnainer metadata key) accessed using the same interface. These accessors could be removed and replaced with direct access to the corresponding data using pointers which can be passed to pkg/cri/nri.API in `Register()`.